### PR TITLE
Add block/unblock/remove APIs.

### PR DESCRIPTION
**Pull request by rajnesh-kanwal**

### CODE REVIEW

File: specification/sbi_atee.adoc
Comment (rsahita) [inline specification/sbi_atee.adoc:1]: general comment - the "block" terminology may be confusing - I have been using "invalidate page mapping" and "validate page mapping" terms - perhaps we can use the same for these ABI functions.

File: specification/sbi_atee.adoc [hunk 1]
```diff
@@ -260,15 +260,26 @@ is that the host will service a subsequent page-fault that results from a TVM-ac
...
-to complete the assignment change. The calling TVM vCPU is considered blocked until the assignment change is completed.
-Attempts to run it with `sbi_tee_host_run_tvm_vcpu()` will fail. Any guest page faults taken by other TVM vCPUs in this
-region before the completion of the assignment change are considered fatal. The host may not insert any pages in the region
-before the completion of the assignment change. Upon completion, the host may either immediately add non-confidential pages
-or may insert the pages on faults during TVM access into the region using `sbi_tee_host_add_tvm_shared_pages()`.
+results in an exit to the TSM which enforces the security properties on the mapping and exits to the VMM host.
+If the region of address space is populated, the host must first invalidate and remove the confidential pages.
+This requires the host to make three separate ECALLs to the TSM:
+ 
+. `sbi_tee_host_tvm_block_pages()`
```
Comment (rsahita) [inline specification/sbi_atee.adoc:267]:
```suggestion
. `sbi_tee_host_tvm_invalidate_page_maps()`
```

File: specification/sbi_atee.adoc [hunk 2]
```diff
@@ -938,6 +949,101 @@ The possible error codes returned in `sbiret.error` are shown below.
 | SBI_ERR_FAILED          | The operation failed for unknown reasons.
 |===
 
+[#sbi_tee_host_tvm_block_pages]
```
Comment (rsahita) [inline specification/sbi_atee.adoc:952]:
```suggestion
[#sbi_tee_host_tvm_invalidate_page_maps]
```

File: specification/sbi_atee.adoc [hunk 3]
```diff
@@ -938,6 +949,101 @@ The possible error codes returned in `sbiret.error` are shown below.
...
+=== Function: TEE Host TVM Block Pages (FID #16)
```
Comment (rsahita) [inline specification/sbi_atee.adoc:953]:
```suggestion
=== Function: TEE Host TVM Invalidate Page Mappings (FID #16)
```

File: specification/sbi_atee.adoc [hunk 4]
```diff
@@ -938,6 +949,101 @@ The possible error codes returned in `sbiret.error` are shown below.
...
+[source, C]
+-----
+struct sbiret sbi_tee_host_tvm_block_pages(unsigned long tvm_guest_id,
```
Comment (rsahita) [inline specification/sbi_atee.adoc:956]:
```suggestion
struct sbiret sbi_tee_host_tvm_invalidate_page_maps(unsigned long tvm_guest_id,
```

File: specification/sbi_atee.adoc [hunk 5]
```diff
@@ -938,6 +949,101 @@ The possible error codes returned in `sbiret.error` are shown below.
...
+                                           unsigned long gpa,
+                                           unsigned long length);
+-----
+
+Invalidates the pages in the specified range of guest physical address space and
```
Comment (rsahita) [inline specification/sbi_atee.adoc:961]:
```suggestion
Invalidates the page mappings in the specified range of guest physical address space and
```

File: specification/sbi_atee.adoc [hunk 6]
```diff
@@ -938,6 +949,101 @@ The possible error codes returned in `sbiret.error` are shown below.
...
+marks the pages as blocked.
```
Comment (rsahita) [inline specification/sbi_atee.adoc:962]:
```suggestion
thus marks the pages as blocked from creating new cached mappings.
```

File: specification/sbi_atee.adoc [hunk 7]
```diff
@@ -938,6 +949,101 @@ The possible error codes returned in `sbiret.error` are shown below.
...
+
+For each page in the range, the TSM must verify that:
+
+* The page is currently marked present in the TVM’s page table.
+* The page is either mapped and uniquely owned by the TVM, or shared and owned by the host.
+
+After verifying these pre-conditions are met, the TSM then invalidates the pages and
+marks the pages as blocked. The host must complete a TVM TLB invalidation sequence,
```
Comment (rsahita) [inline specification/sbi_atee.adoc:970]:
```suggestion
marks the pages as not-present. The host must complete a TVM TLB invalidation sequence,
```

File: specification/sbi_atee.adoc [hunk 8]
```diff
@@ -938,6 +949,101 @@ The possible error codes returned in `sbiret.error` are shown below.
...
+initiated by `sbi_tee_host_tvm_fence()`, in order to complete the invalidation.
+
+Guest page faults taken by the TVM on blocked pages continue to be reported to the host.
```
Comment (rsahita) [inline specification/sbi_atee.adoc:973]:
```suggestion
Guest page faults taken by the TVM on invalid page mappings continue to be reported to the host.
```

File: specification/sbi_atee.adoc [hunk 9]
```diff
@@ -938,6 +949,101 @@ The possible error codes returned in `sbiret.error` are shown below.
...
+The page remains invalid until the mapping is unblocked (marked present) removed, or
```
Comment (rsahita) [inline specification/sbi_atee.adoc:974]:
```suggestion
The page mapping remains invalid until the mapping is unblocked (marked present) removed, or
```

File: specification/sbi_atee.adoc [hunk 10]
```diff
@@ -938,6 +949,101 @@ The possible error codes returned in `sbiret.error` are shown below.
...
+becomes part of a huge page by promotion/demotion operation.
+
+The possible error codes returned in `sbiret.error` are shown below.
+
+[#table_sbi_tee_host_tvm_block_pages_errors]
```
Comment (rsahita) [inline specification/sbi_atee.adoc:979]:
```suggestion
[#table_sbi_tee_host_tvm_invalidate_page_maps_errors]
```

File: specification/sbi_atee.adoc [hunk 11]
```diff
@@ -938,6 +949,101 @@ The possible error codes returned in `sbiret.error` are shown below.
...
+.TEE Host TVM Block Pages
```
Comment (rsahita) [inline specification/sbi_atee.adoc:980]:
```suggestion
.TEE Host TVM Invalidate Page Mappings
```

File: specification/sbi_atee.adoc [hunk 12]
```diff
@@ -938,6 +949,101 @@ The possible error codes returned in `sbiret.error` are shown below.
...
+[cols="2,3", width=90%, align="center", options="header"]
+|===
+| Error code              | Description
+| SBI_SUCCESS             | The operation completed successfully.
+| SBI_ERR_INVALID_PARAM   | `tvm_guest_id` or `length` were invalid.
+| SBI_ERR_INVALID_ADDRESS | `gpa` was invalid.
+| SBI_ERR_FAILED          | The operation failed for unknown reasons.
+|===
+
+[#sbi_tee_host_tvm_unblock_pages]
```
Comment (rsahita) [inline specification/sbi_atee.adoc:990]: can you please make similar changes for the unblock_pages -> validate_page_maps

**Review by rsahita — CHANGES_REQUESTED** [review]
LGTM - only change requested is to use the term "Invalidate page mappings" instead of "block pages" (and vice-versa).

File: specification/sbi_atee.adoc
Comment (rajnesh-kanwal) [reply specification/sbi_atee.adoc:1]: Adding @abrestic-rivos to provide his input, as he designed these APIs.

Comment (abrestic-rivos) [reply specification/sbi_atee.adoc:1]: I'm fine with the rename. My only complaint with the "invalidate/validate page mappings" is it's a bit verbose. How about `{in,}validate_ptes` or `{in,}validate_pages` or `set_ptes_{present,invalid}`, etc.

Comment (rajnesh-kanwal) [reply specification/sbi_atee.adoc:1]: `{in,}validate_pages` SGTM.

**Review by rsahita — APPROVED** [review]

**Pull request merged**